From d880b794f29d190fb3ee9e6260b3cfb9b3d225a0 Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Thu, 25 Sep 2025 17:03:37 +0200
Subject: [PATCH 926/954] feat(clk): introduce support of fixed clock

Add fixed clock management support to clock framework.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git a9641a35292022bd6483e65d2ce16fb56ab6a0c7 feat(clk): introduce support of fixed clock]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I25a68fc613fb017136ad2624dc93c26e9ec2257d
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/493937
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Maxime MERE <maxime.mere@foss.st.com>
---
 drivers/clk/clk-fixed.c     | 125 ++++++++++++++++++++++++++++++++++++
 include/drivers/clk-fixed.h |  12 ++++
 plat/st/common/common.mk    |   1 +
 3 files changed, 138 insertions(+)
 create mode 100644 drivers/clk/clk-fixed.c
 create mode 100644 include/drivers/clk-fixed.h

diff --git a/drivers/clk/clk-fixed.c b/drivers/clk/clk-fixed.c
new file mode 100644
index 0000000000..f2f12eef4a
--- /dev/null
+++ b/drivers/clk/clk-fixed.c
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2025, STMicroelectronics - All Rights Reserved
+ * Author(s): Gabriel Fernandez, <gabriel.fernandez@st.com> for STMicroelectronics.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <lib/utils_def.h>
+
+#include <common/debug.h>
+#include <common/fdt_wrappers.h>
+#include <drivers/clk.h>
+
+#include <platform_def.h>
+
+#ifdef CFG_CLK_FIXED_NB
+#define MAX_FIXED_CLOCK_PROVIDERS CFG_CLK_FIXED_NB
+#else
+#define MAX_FIXED_CLOCK_PROVIDERS 10U
+#endif
+
+struct clock_fixed_provider_t {
+	unsigned long rate;
+	uint16_t phandle;
+};
+
+static struct clock_fixed_providers_t {
+	struct clock_fixed_provider_t providers[MAX_FIXED_CLOCK_PROVIDERS];
+	uint8_t nb;
+} clk_fixed_providers;
+
+static int clk_fixed_enable(unsigned long binding_id)
+{
+	/* Fixed clock is always enabled */
+	return 0;
+}
+
+static void clk_fixed_disable(unsigned long binding_id)
+{
+	/* Fixed clocks are always enabled, so no action needed. */
+}
+
+static bool clk_fixed_is_enabled(unsigned long binding_id)
+{
+	/* Fixed clocks are always enabled. */
+	return true;
+}
+
+static unsigned long clk_fixed_get_rate(unsigned long binding_id)
+{
+	/* Binding id is the phandle */
+	uint16_t phandle = binding_id;
+	uint8_t i;
+
+	for (i = 0U; i < clk_fixed_providers.nb; i++) {
+		if (clk_fixed_providers.providers[i].phandle == phandle) {
+			return clk_fixed_providers.providers[i].rate;
+		}
+	}
+
+	return 0UL;
+}
+
+static int clk_fixed_get_parent(unsigned long binding_id)
+{
+	return -1;
+}
+
+const struct clk_ops fixed_clock_ops = {
+	.enable		= clk_fixed_enable,
+	.disable	= clk_fixed_disable,
+	.is_enabled	= clk_fixed_is_enabled,
+	.get_rate	= clk_fixed_get_rate,
+	.get_parent	= clk_fixed_get_parent,
+};
+
+static void clk_add_fixed_clock(uint32_t phandle, unsigned long frequency)
+{
+	if (clk_fixed_providers.nb >= MAX_FIXED_CLOCK_PROVIDERS) {
+		EARLY_ERROR("No more room for fixed clock providers\n");
+		panic();
+	}
+
+	clk_fixed_providers.providers[clk_fixed_providers.nb].phandle = phandle;
+	clk_fixed_providers.providers[clk_fixed_providers.nb].rate = frequency;
+	clk_fixed_providers.nb++;
+
+	clk_add_provider(phandle, &fixed_clock_ops);
+}
+
+void clk_fixed_register(void *fdt)
+{
+	int node = 0;
+	int subnode = 0;
+	unsigned long frequency = 0UL;
+
+	node = fdt_path_offset(fdt, "/clocks");
+	if (node < 0) {
+		return;
+	}
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		const fdt32_t *cuint = NULL;
+
+		if (!fdt_node_is_enabled(fdt, subnode)) {
+			continue;
+		}
+
+		if (fdt_node_check_compatible(fdt, subnode, "fixed-clock")) {
+			continue;
+		}
+
+		cuint = fdt_getprop(fdt, subnode, "clock-frequency", NULL);
+		if (cuint == NULL) {
+			frequency = 0UL;
+		} else {
+			frequency = fdt32_to_cpu(*cuint);
+		}
+
+		clk_add_fixed_clock(subnode, frequency);
+	}
+}
diff --git a/include/drivers/clk-fixed.h b/include/drivers/clk-fixed.h
new file mode 100644
index 0000000000..8ffd5b7b22
--- /dev/null
+++ b/include/drivers/clk-fixed.h
@@ -0,0 +1,12 @@
+/*
+ * Copyright (c) 2025, STMicroelectronics - All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef CLK_FIXED_H
+#define CLK_FIXED_H
+
+void clk_fixed_register(void *fdt);
+
+#endif /* CLK_FIXED_H */
diff --git a/plat/st/common/common.mk b/plat/st/common/common.mk
index f7a5d4217e..a2e9272840 100644
--- a/plat/st/common/common.mk
+++ b/plat/st/common/common.mk
@@ -182,6 +182,7 @@ include lib/xlat_tables_v2/xlat_tables.mk
 PLAT_BL_COMMON_SOURCES		+=	${XLAT_TABLES_LIB_SRCS}
 
 PLAT_BL_COMMON_SOURCES		+=	drivers/clk/clk.c				\
+					drivers/clk/clk-fixed.c		\
 					drivers/delay_timer/delay_timer.c		\
 					drivers/delay_timer/generic_delay_timer.c	\
 					drivers/st/clk/stm32mp_clkfunc.c		\
-- 
2.53.0

